### ISSUE TITLE
FFT precision appears to ignore jax.default_matmul_precision setting

### ISSUE THREAD

**Issue by WillFarris** [google/jax#9881]
For a university project I'm trying to implement a Mersenne prime search program using the TPU as a hardware accelerator. The algorithm we're using uses the FFT to square numbers mod the Mersenne prime candidate, and this calculation is quite prone to precision errors. We've discovered that `bfloat16` is too imprecise for the calculations we're trying to perform and thus will need to see if `bfloat16_3x` will give us the precision we need.

In some JAX functions such as `jax.numpy.matmul()` I can manually specify a `precision` keyword argument to tell JAX to use a specific precision for the operation. This is apparently not an option with `jax.numpy.fft.fft()` so we have resulted to using the `jax.default_matmul_precision` config value to attempt to force the FFT to use the highest precision possible on the TPU. However, even with this flag set to `HIGHEST` we are still getting the exact same precision error whether `HIGHEST` is set or not.

So my question is, how do we force the `jax.numpy.fft()` function to use `bfloat16_3x`/`tensorfloat32` for its internal calculations? Or is there a way to verify the internal precision of the calculations done on the TPU? If specifying the precision is not possible for the FFT, could this be clarified in the documentation so others are aware of the issue?

Below is the code that will reproduce the issue, and a few sample precision errors. This code was run in Google Colab:

```
import jax
import jax.numpy as jnp
from jax import jit, lax, device_put
from functools import partial

import jax.tools.colab_tpu
jax.tools.colab_tpu.setup_tpu()

# Tell JAX to use the highest possible precision by default
jax.config.update("jax_default_matmul_precision", jax.lax.Precision.HIGHEST)

# Example of squaring a number using the FFT
x = jnp.array([6561, 0, 0, 0], dtype=jnp.float32)
x = jnp.fft.fft(x)
x = jnp.multiply(x, x)

# On a system with zero precision error, x is an integer with no imaginary component
x = jnp.real(jnp.fft.ifft(x))

# We compute the difference between the actual result and the next nearest integer (this is our precision error)
x_rounded = jnp.round(x)
print(x - x_rounded)
```

- TPU roundoff error, default precision: [ 0.0000000e+00  1.3183920e-07  0.0000000e+00 -1.8457621e-08]
- TPU roundoff error, highest precision: [ 0.0000000e+00  1.3183920e-07  0.0000000e+00 -1.8457621e-08]
-  GPU roundoff error, default precision: [0. 0. 0. 0.]
( note the roundoff error is *exactly* the same for both the default and highest default precision setting on the TPU)

**Comment by shoyer** [comment]
My understanding is that the JAX FFT implementation on TPU always uses HIGHEST precision matrix multiplication internally, which explains why you haven't found an option to customize it.

If it isn't high enough precision, you may need to use CPU or GPU implementations for now, particularly if you need float64 precision.

**Comment by shoyer** [comment]
I think we would probably be OK to add a clarifying note about precision support to jax.numpy.fft documentation, to clarify that matrix-multiplication precision flags do not affect the accuracy of these functions, even if they use matrix-multiplication internally. Though I hope that would also be your first guess, given we do not have any options for customizing precision.
